timemachine/fe/free_energy.py

```diff
@@ -725,3 +722,184 @@ def compute_intermediate_result(lambdas: Sequence[float]) -> PairBarResult:
     boxes = [get_state(lamb).boxes for lamb in lambdas]
 
     return results, frames, boxes
+
+
+def run_sims_hrex(
+    initial_states: Sequence[InitialState],
+    md_params: MDParams,
+    n_frames_per_iter: int,
+    temperature: float,
+    n_swap_attempts_per_iter: Optional[int] = None,
+    verbose: bool = True,
+) -> Tuple[PairBarResult, List[StoredArrays], List[NDArray], HrexDiagnostics]:
+    r"""Sample from a sequence of states using nearest-neighbor Hamiltonian Replica EXchange (HREX).
+
+    See documentation for :py:func:`timemachine.md.hrex.run_hrex` for details of the algorithm and implementation.
+
+    Parameters
+    ----------
+    initial_states: sequence of InitialState
+        States to sample
+
+    md_params: MDParams
+        Parameters used to simulate new states
+
+    n_frames_per_iter: int
+        Number of frames to sample using MD per iteration
+
+    temperature: float
+        Temperature in K
+
+    n_swap_attempts_per_iter: int or None, optional
+        Number of nearest-neighbor swaps to attempt per iteration. Defaults to len(initial_states) ** 4.
+
+    verbose: bool, optional
+        Whether to print diagnostic information
+
+    Returns
+    -------
+    PairBarResult
+        results of pair BAR free energy analysis
+
+    List[StoredArrays]
+        coord trajectories
+
+    List[NDArray]
+        box trajectories
+
+    HrexDiagnostics
+        HREX statistics (e.g. swap rates, replica-state distribution)
+    """
+
+    if n_swap_attempts_per_iter is None:
+        n_swap_attempts_per_iter = get_swap_attempts_per_iter_heuristic(len(initial_states))
+
+    np.random.seed(md_params.seed)
```

Comment:
Agreed on the preference for no side effects. Unfortunately I think this is required for now because `timemachine.md.moves.MonteCarloMove` relies on the global numpy RNG state. This could potentially be refactored to accept a seed or generator argument, but this seemed like a task for a separate PR. We might track this in #980.

Reply:
I have a preference for doing this now, as this just seems to risk changing behavior that we don't intend to. If is to be done in a separate PR I would say that this shouldn't be used in the `run_leg` methods until we remove the side effect.

Reply:
>Unfortunately I think this is required for now because timemachine.md.moves.MonteCarloMove relies on the global numpy RNG state. This could potentially be refactored to accept a seed or generator argument, but this seemed like a task for a separate PR.

The dependence on global numpy RNG state in `MonteCarloMove` was not well-considered and is not something I think we need to preserve.

Including a seed in the constructor `HREX(..., seed=1234)` ... `HREX::move(x) = ... behavior that depends on and updates self.rng state ...` may be one way to allow to seed the move without depending on / modifying global numpy RNG state and without refactoring / modifying signatures anywhere.

There is not currently a fully consistent convention in timemachine, and I think all 3 approaches are used in various places (global rng state, instance rng state, functions accept rng state as input).

Reply:
> without refactoring / modifying signatures anywhere.

If I'm not mistaken, this would still require updating `MonteCarloMove` and subclasses, correct? (Unless we remove the inheritance for now and just have `NeighborSwapMove` reimplement `move`)

Reply:
Ahh, sorry, yes -- this would require changing a few more places than I realized.

Would require:
* modifying signatures and implementations of constructors (to accept `seed=1234` and add `self.rng = np.random.default_rng(seed)`)
* modifying the use of the unif(0,1) r.v. for acceptance test https://github.com/proteneer/timemachine/blob/3984fe517d92904986614515810ba392dc60f5a4/timemachine/md/moves.py#L51-L53 (to use `self.rng.uniform(0,1)` instead of `np.random.rand()`)

Just wouldn't require modifying places that call `.move(x)`.

Changing the places that call `.move(x)` to be `.move(x, key)` or similar may be the about the same difficulty.

>(Unless we remove the inheritance for now and just have NeighborSwapMove reimplement move)

That also seems fine to me -- the only things MonteCarloMove defines are simple, generic, and easily replicated (like an acceptance counter)

Reply:
After some time attempting various fixes, I wasn't able to find a satisfactory solution to removing the dependence on global random state that seemed reasonable to implement in this PR. The approaches I tried were:

1. **Add random state as instance variable to `Move` and subclasses, add `seed` to constructor.** this becomes quite tedious (we have to initialize each neighbor swap move and the mixture move with a random seed and ensure that they are all different) and I found that it obscures the mathematical definitions
1. **Augment signature of `propose` and `move` methods to accept a `numpy.random.Generator` instance.** This seems to have better compositionality than (1), e.g. it's not necessary to initialize separate PRNGs for each move in a mixture, but is incompatible with `DeterministicMTMMove`, which uses the JAX random API.
1. **Augment signature of `propose` and `move` methods to accept a `jax.random.PRNGKeyArray`. This approach has the compositionality of (2) and works with the existing `DeterministicMTMMove`, but generating random numbers with JAX is *much* [slower](https://github.com/google/jax/issues/968) than with numpy unless we're able to use `jax.jit` (in which case it's blazingly fast). This would probably be my preferred approach in the future, but it would require some refactoring (I think mainly just in the HREX code) due to the restriction that `State` must be a PyTree (e.g. immutable collection of primitive types or `jax.Array`s)

For this PR, I would propose to remove the call to `np.random.seed` in `run_sims_hrex` and rely on the caller to set the numpy random state (with the understanding that this is _not_ controlled by `md_params.seed`, which I will emphasize in the docstring). I think this aligns well with existing convention (i.e. places where we currently use `MonteCarloMove`), and will allow us time to design a good solution to PRNG state. @badisa @maxentile what do you think?

Reply:
Thanks for investigating and reporting on these.

Any of these 3 described refactors sounds appropriate to me for later, and for this PR either the currently implemented approach (which resets global numpy random state within `run_sims_hrex`) or the proposed approach (which has `run_sims_hrex` depend on and modify global numpy random state) sounds appropriate.

Reply:
In the end, I opted to keep setting the numpy random state in `run_sims_hrex`, but added a warning and todo to address this in https://github.com/proteneer/timemachine/pull/1128/commits/2db9fd794d8a434f3ad8ef37945ca3c8f98c235f